File: Circuit_Detector/segment.py
import cv2
import numpy as np

from . import editor
from .opjects import *
from .src.LoodDetect import CDEC
from .constant import *

my_opjects = [Resistance(), DCS(), Inductor(), Capac(), ACS(), Gnd()]
#my_opjects = [Resistance()]
NOT_CIR_COMP = [Gnd]

def FixNoding(nodes, circuit):
	cnt = 1
	tmp = {}
	rem_dict = []
	gnd = []
	
	for key, val in nodes.items():
		for comp in circuit:
			if comp.__class__ == Gnd:
				if val in comp.nodes:
					gnd.append(val)

	for key, val in nodes.items():
		keep = 0
		if val in tmp:
			continue
		if val in gnd:
			tmp[val], nodes[key] = 0, 0
			continue
		for comp in circuit:
			if val == comp.n1 or val == comp.n2:
				tmp[val], nodes[key] = cnt, cnt
				cnt += 1
				keep = 1
				break
		if not keep:
			rem_dict.append(key)
	for key in rem_dict:
		del nodes[key]

	for comp in circuit:
		if comp.__class__ == Gnd:
			continue
		if comp.n1 in tmp:
			comp.n1 = tmp[comp.n1]
		if comp.n2 in tmp:
			comp.n2 = tmp[comp.n2]
		if comp.n1:
			comp.n1 = f'n{comp.n1}'
		if comp.n2:
			comp.n2 = f'n{comp.n2}'
		
def CreatCircuit(gray, nodes, node_map, components):
	circuit = []

	for opj in my_opjects:
		opj.__class__.index = 1

	for opj in components:
		if opj.__class__ == ACS:
			opj.index = DCS.index
			DCS.index += 1
		else:
			opj.index = opj.__class__.index
			opj.__class__.index += 1
		opj.getNode(gray, nodes, node_map)
		circuit.append(opj)

	FixNoding(nodes, circuit)
	for opj in circuit:
		if opj.__class__ in NOT_CIR_COMP:
			circuit.remove(opj)
	return circuit

def simplfySkel(skel, state=1):
	n, m = skel.shape
	vis = np.zeros((n, m), dtype=int) 
	result = {}

	dic = CDEC.SimplfySkel(skel, vis, n, m, state)
	
	for i in range(1, dic[0].key):
		result[dic[i].key] = dic[i].val

	CDEC.free_dict(dic)
	return ([result, vis])

def IsIntersection(comp, op):
	Inter = editor.get2dInersection(comp.shape, op.shape)
	if Inter[0][0] != -1:
		if comp.score >= op.score:
			op = comp
		op.shape = Inter[1]
		return (op)	
	return (0)

def AddComponent(components, op):
	going = 1

	while going:
		going = 0
		for comp in components:
			iscomp = IsIntersection(comp, op)
			if iscomp:
				op, going = iscomp, 1
				components.remove(comp)
	components.append(op)

def getComponents(src):
	skel = src.copy()
	components = []
	opj = []

	for comp in my_opjects:
		opj.extend(comp.detect(skel))
	for op in opj:
		AddComponent(components, op)
	for comp in components:
		x, y, w, h = comp.shape
		editor.remove_part(skel, x + 1, y + 1, w - 2, h - 2)

	return [skel, components]

File: Circuit_Detector/src/segment.c
#include "Detect.h"


int* CompSize(unsigned char *img, int rows, int cols, int x, int y, int w, int h) {
    int w1 = 2400, w2 = -1, h1 = 2400, h2 = -1;

    for (int i = x; i < x + w; ++i) {
        for (int j = y; j < y + h; ++j) {
            if (img[i * cols + j]) {
                w1 = MIN(w1, j);
                w2 = MAX(w2, j);
                h1 = MIN(h1, i);
                h2 = MAX(h2, i);
            }
        }
    }

    int* result = (int*)malloc(4 * sizeof(int));

    result[1] = w1;
    result[0] = h1;
    result[3] = w2 - w1;
    result[2] = h2 - h1;

    return (result);
}

int *nodeScale(int *nodes, int *vis, int rows, int cols, int n){
    int index;
    int *res = malloc(sizeof(int) * 4);
    res[0] = res[1] = 2400;
    res[2] = res[3] = 0;
    for (int i = 0; i < rows; i++){
        for (int j = 0; j < cols; j++){
            index = i * cols + j;
            if (vis[index])
                for (int k = 0; k < n; k++)
                    if (vis[index] == nodes[k]){
                        res[0] = MIN(res[0], j);
                        res[2] = MAX(res[2], j);
                        res[1] = MIN(res[1], i);
                        res[3] = MAX(res[3], i);
                        break;
                    }
        }
    }
    res[2] -= res[0];
    res[3] -= res[1];
    return (res);
}

int CirDfs(unsigned char *skel, int *vis, int rows, int cols, int l, int r, int val, int deep){
    if (l < 0 || l >= rows || r < 0 || r >= cols)
        return (0);

    int index = l * cols + r;
    if (vis[index] || !skel[index]) return (0);
    
    int len = 1;
    vis[index] = val;    

    if (deep > max_lenght) return 1;

    for (int i = l - boxwire; i <= l + boxwire; i++)
        for (int j = r - boxwire; j <= r + boxwire; j++)
            len += CirDfs(skel, vis, rows, cols, i, j, val, deep + 1);

    return (len);
}

Dict *SimplfySkel(unsigned char *img, int *vis, int rows, int cols, int state){
    int node_size, node, cnt, index, isit;
    int len;
    Dict *res;

    node_size = 1;
    node = cnt = 1;

    res = (Dict*)malloc(node_size * sizeof(Dict));
    res[0].key = node_size;

    for (int i = 0; i < rows; i++){
        for (int j = 0; j < cols; j++){
            index = i * cols + j;
            
            if (!img[index]) continue;

            if (vis[index]){
                isit = 0;
                for (int k = 1; k < node_size; k++)
                    if (vis[index] == res[k].key){
                        isit = 1;
                        break;
                    }
                img[index] *= isit;   
                continue;
            }

            len = CirDfs(img, vis, rows, cols, i, j, cnt, 1);
            if (len >= wire_lenght * state && len <= max_lenght){
                node_size++;
                res = (Dict*)realloc(res, node_size * sizeof(Dict));
                res[0].key = node_size;

                res[node_size - 1].key = cnt;
                res[node_size - 1].val = node;
                node++;
            }else img[index] = 0;
            cnt++;
        }
    }  
    return (res);
}
